README.md
osm_akuechler [![Build Status](https://travis-ci.org/admiralsmaster/osm_akuechler.svg?branch=master)](https://travis-ci.org/admiralsmaster/osm_akuechler) [![Codacy Badge](https://api.codacy.com/project/badge/grade/ca55f3e5de3548ffbfb64a63cc1adf93)](https://www.codacy.com/app/github-ariel/osm_akuechler)
=================================================================

A Joomla Plugin for Joomla 2.5 and 3.x to show an Openstreetmap map.
The plugin use the Leaflet javascript library.

1. [Joomla](http://www.joomla.org/)
2. [Open Street Map](http://www.openstreetmap.org/)
3. [Leaflet](http://leafletjs.com/)

The advantage of OpenStreetMap is that no license fees such as Google maps (see [here](https://developers.google.com/maps/faq#usage_calculated)) are necessary. OpenStreetMap (OSM) is a collaborative project to create a free editable map of the world. (see [Wikipedia](https://en.wikipedia.org/wiki/OpenStreetMap))

Example
----------------

An example can be found [here](https://www.kuechler.info/osm).

Usage
---------------

Write in one line, i.e.

```xml
<div class="osm-map" data-lat="51.047382" data-lon="13.734087" data-zoom="17">
  <div class="osm-point" data-lat="51.047382" data-lon="13.734087">
    <h3>
    	Spielplatz in der Innenstadt von Dresden.
  	</h3>
  </div>
  <div class="osm-point" data-lat="51.0466" data-lon="13.733087">
    <h3>
    	Parkplatz.
  	</h3>
    <p>
      Hier könnt ihr parken.
    </p>
  </div>
</div>
```

Name	| Description
------- | -------------
osm-map 	| DIV tag with the card data.
osm-point	| DIV tag with a points, you can add lots of points. Included HTML text is for marker popup.
data-lat 	| Coordinate, Latitude
data-lon 	| Coordinate, Longitude
data-zoom 	| Zoom factor, depends on tiles server, between 1 and 18.
height 	| The height of DIV is the height of the card. Set the height via CSS, default is 400px.


CHANGELOG
--------------
 see [CHANGELOG.md](CHANGELOG.md)

License
------------------

You can choose: [MIT License](http://opensource.org/licenses/mit-license) or [GPL-3.0](http://opensource.org/licenses/gpl-3.0)

Leaflet is provided with the [BSD 2-Clause License](https://github.com/Leaflet/Leaflet/blob/master/LICENSE).


Configuration
-------------------

No idea? Use the default value.

You can configurate the tile server URL. A tile server provides up to date map images. The default value is http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png. Another tile server is i.e. the Wikipedia tile server in Germany: http://{s}.www.toolserver.org/tiles/germany/{z}/{x}/{y}.png. Or you can choose you own tile server. See OpenStreetMap Wiki [here](http://wiki.openstreetmap.org/wiki/TMS) and [here](http://wiki.openstreetmap.org/wiki/Tileserver). The describtion of the placeholders {s}, {x}, {y}, {z} can you find in the Leaflet [documentation](http://leafletjs.com/reference.html#url-template).

You can choose that the leaflet library is loaded not from your server but from Cloudflares [CdnJS](http://cdnjs.com/). This CDN (see [Wikipedia](https://en.wikipedia.org/wiki/Content_delivery_network)) can rise the perfomance of your page and can reduce your webservers load.



CHANGELOG.md

* **Version 2.10.0**: (June 13nd, 2019) update leafletjs to 1.5.1 #10
* **Version 2.9.0**: (April 30nd, 2019) update leafletjs to 1.4.0 #9
* **Version 2.8.0**: (August 22nd, 2018) update leafletjs to 1.3.4 #7
* **Version 2.7.0**: (April 10th, 2018) update leafletjs to 1.3.0 #6
* **Version 2.6.0**: (October 9th, 2017) update leafletjs to 1.2.0 #3, add update server, fixes #1 #2
* **Version 2.5.0**: (June 28th, 2017) update leafletjs to 1.1.0, add [unpkg.com](https://unpkg.com/) CDN
* **Version 2.4.0**: (April 29th, 2017) update leafletjs to 1.0.3
* **Version 2.3.0**: (Oct 4th, 2016) update leafletjs to 1.0.1
* **Version 2.2.0**: (Sep 27, 2016) update leafletjs to 1.0.0
* **Version 2.1.0**: (Dec 1, 2015) update leafletjs to 0.7.7
* **Version 2.0.0**: (May 13, 2015) Change the API. Now you can use HTML-Tags only.
* **Version 1.3.0**: (November 25, 2014) add support for [jsdelivr](http://www.jsdelivr.com/)
* **Version 1.2.0**: (July 27, 2014) leafletjs upgrade to 0.7.3, Support Interaction Options from leaflet, Fix default value for retina support
* **Version 1.1.0**: (January 12, 2014) add retina configuration, possiblility to configure max zoom for tiles and copyright hint.
* **Version 1.0.6**: (January 12, 2014) leaflet 0.7.1
* **Version 1.0.5**: (July 11, 2013) leaflet 0.6.2, possible to use Cloudflares [CdnJS](http://cdnjs.com/)
* **Version 1.0.4**: (November 05, 2012) IE CSS usage, not running on admin pages, no backslash at Windows system, better code
* **Version 1.0.3**: (Oct. 26, 2012) leaflet 0.4.5
* **Version 1.0.2**: Bugfix "+" to "." String concat
* **Version 1.0.1**:
* **Version 1.0.0**: initial version, leaflet 0.4


js/Osm_akuechler.js
jQuery(document).ready(function() {
    jQuery(".osm-map").each(function() {
        var map = jQuery(this),
            lat = map.attr("data-lat"),
            lon = map.attr("data-lon"),
            zoom = map.attr("data-zoom"),
            id = map.attr("id"),
            js;
        
        if (id === undefined) {
            id = "map" + Math.floor((1 + Math.random()) * 0x10000000).toString(16).substring(1);
            map.attr("id", id);
        }

        js = "{\n"
           + "var map=new L.Map('" + id + "', osmAKuechlerConfig.mapConfig);\n"
           + "map.attributionControl.setPrefix('');\n"
           + "var baselayer=new L.TileLayer(osmAKuechlerConfig.server, osmAKuechlerConfig.tileConfig);\n"
           + "var koord=new L.LatLng(" + lat + ", " + lon + ");\n"
           + "map.setView(koord, " + zoom + ").addLayer(baselayer);\n"
           + "\n";

        map.children(".osm-point").each(function() {
            var point = jQuery(this),
                lat = point.attr("data-lat"),
                lon = point.attr("data-lon"),        
                data = point.html();

            js += "{\n"
               + "var koord=new L.LatLng(" + lat + ", " + lon + ");\n"
               + "var marker=new L.Marker(koord);\n"
               + "map.addLayer(marker);\n"
               + "marker.bindPopup('" + data.replace(/[\n\r]+/g, "") + "');\n"
               + "}\n";

            point.remove();
        });
        js += "}\n";
        
        map.append("<script>" + js + "</script>");
    });
});


Osm_akuechler.php
<?php
/**
 * @package      Joomla
 * @copyright    Copyright (C) 2011-2019 Ariel Küchler. All rights reserved.
 * @license      MIT License (http://opensource.org/licenses/mit-license copyright information see above) OR GPL-3.0 (http://opensource.org/licenses/gpl-3.0)
 */
defined ( '_JEXEC' ) or die ( 'Restricted access' );

JHtml::_ ( 'jquery.framework' );

jimport ( 'joomla.plugin.plugin' );
class plgContentOsm_akuechler extends JPlugin {
    var $copyright = '&copy; <a href="http://www.openstreetmap.org" target="_blank">OpenStreetMap</a> contributors, <a href="http://creativecommons.org/licenses/by-sa/2.0/" target="_blank">CC BY-SA</a>';
    var $tileServer = 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png';
    var $maxZoom = 18;
    var $cdnUsage = 0;
    var $detectRetina = true;
    var $dragging = true;
    var $touchZoom = true;
    var $scrollWheelZoom = true;
    var $doubleClickZoom = true;
    var $boxZoom = true;
    var $tap = true;
    var $tapTolerance = 15;
    var $trackResize = true;
    var $worldCopyJump = false;
    var $closePopupOnClick = true;
    var $bounceAtZoomLimits = true;
    var $version = '@leaflet-version@';
       
    function __construct(&$subject, $config) {
        parent::__construct ( $subject, $config );
    }

    function onContentPrepare($context, &$row, &$params, $limitstart = 0) {
        // fast fail
        $app = JFactory::getApplication ();
        if ($app->isAdmin () || (JString::strpos ( $row->text, '{osm' ) === false && JString::strpos ( $row->text, 'osm-map' ) === false)) {
            return true;
        }
        
        // old
        $regexOld = '/\{osm\s+(.*?)\}/i';
        // find all instances of plugin and put in $matches
        preg_match_all ( $regexOld, $row->text, $matches, PREG_PATTERN_ORDER );
        // Number of plugins
        $countOld = count ( $matches [1] );
        
        // new
        $regexNew = "/class\s*=\s*[\"']?[^\"']*osm-map[\"'\s>]/i";
        $resultNew = preg_match_all ( $regexNew, $row->text );
        
        // plugin only processes if there are any instances of the plugin in the text
        if ($countOld || $resultNew) {
            $this->_addConfig ();
            $this->_addLeafletScripts ();
            
            // old
            for($i = 0; $i < $countOld; $i ++) {
                $row->text = str_replace ( $matches [0] [$i], $this->_getReplacmentOld ( $matches [1] [$i] ), $row->text );
            }
            // new
            if ($resultNew) {
                $document = JFactory::getDocument ();
                $document->addScript ( 'plugins/content/Osm_akuechler/js/Osm_akuechler.min.js' );
                $document->addStyleSheet ( 'plugins/content/Osm_akuechler/css/Osm_akuechler.min.css' );
            }
        }
        return true;
    }

    function _addLeafletScripts() {
        $document = JFactory::getDocument ();
        
        if ($this->params->get ( 'cdn-usage', $this->cdnUsage ) == 1) {
            $document->addStyleSheet ( 'https://cdnjs.cloudflare.com/ajax/libs/leaflet/' . $this->version . '/leaflet.css' );
            $document->addScript ( 'https://cdnjs.cloudflare.com/ajax/libs/leaflet/' . $this->version . '/leaflet.js' );
        } else if ($this->params->get ( 'cdn-usage', $this->cdnUsage ) == 2) {
            $document->addStyleSheet ( 'https://cdn.jsdelivr.net/npm/leaflet@' . $this->version . '/dist/leaflet.css' );
            $document->addScript ( 'https://cdn.jsdelivr.net/npm/leaflet@' . $this->version . '/dist/leaflet.js' );
        } else if ($this->params->get ( 'cdn-usage', $this->cdnUsage ) == 3) {
            $document->addStyleSheet ( 'https://unpkg.com/leaflet@' . $this->version . '/dist/leaflet.css' );
            $document->addScript ( 'https://unpkg.com/leaflet@' . $this->version . '/dist/leaflet.js' );
        } else {
            $document->addStyleSheet ( 'plugins/content/Osm_akuechler/leaflet/leaflet.css' );
            $document->addScript ( 'plugins/content/Osm_akuechler/leaflet/leaflet.js' );
        }
    }

    function _addConfig() {
        $document = JFactory::getDocument ();
        
        $copyright = $this->params->get ( 'copyright', $this->copyright );
        $server = $this->params->get ( 'url', $this->tileServer );
        $maxZoom = $this->params->get ( 'maxZoom', $this->maxZoom );
        $detectRetina = $this->params->get ( 'detect-retina', $this->detectRetina );
        //
        $dragging = $this->params->get ( 'dragging', $this->dragging );
        $touchZoom = $this->params->get ( 'touchZoom', $this->touchZoom );
        $scrollWheelZoom = $this->params->get ( 'scrollWheelZoom', $this->scrollWheelZoom );
        $doubleClickZoom = $this->params->get ( 'doubleClickZoom', $this->doubleClickZoom );
        $boxZoom = $this->params->get ( 'boxZoom', $this->boxZoom );
        $tap = $this->params->get ( 'tap', $this->tap );
        $tapTolerance = $this->params->get ( 'tapTolerance', $this->tapTolerance );
        $trackResize = $this->params->get ( 'trackResize', $this->trackResize );
        $worldCopyJump = $this->params->get ( 'worldCopyJump', $this->worldCopyJump );
        $closePopupOnClick = $this->params->get ( 'closePopupOnClick', $this->closePopupOnClick );
        $bounceAtZoomLimits = $this->params->get ( 'bounceAtZoomLimits', $this->bounceAtZoomLimits );
        
        $document->addScriptDeclaration ( "\n\n// start map config\nvar osmAKuechlerConfig = {\"server\": \"$server\", \"mapConfig\": {\"dragging\": $dragging, \"touchZoom\": $touchZoom, \"scrollWheelZoom\": $scrollWheelZoom, \"doubleClickZoom\": $doubleClickZoom, \"boxZoom\": $boxZoom, \"tap\": $tap, \"tapTolerance\": $tapTolerance,\" trackResize\": $trackResize, \"worldCopyJump\": $worldCopyJump, \"closePopupOnClick\": $closePopupOnClick, \"bounceAtZoomLimits\": $bounceAtZoomLimits}, \"tileConfig\": {\"maxZoom\": $maxZoom, \"attribution\": " . json_encode ( $copyright, JSON_HEX_QUOT ) . ", \"detectRetina\": $detectRetina}};\n// end map config\n\n" );
    }

    function _getReplacmentOld(&$match) {
        $document = JFactory::getDocument ();
        
        $zoom = $this->_getVariableOld ( 'zoom', $match );
        $height = $this->_getVariableOld ( 'height', $match );
        $lat = $this->_getVariableOld ( 'lat', $match );
        $lon = $this->_getVariableOld ( 'lon', $match );
        $popup = $this->_getVariableOld ( 'popup', $match );
        
        $uuid = uniqid ( '', false );
        
        $j = "\n";
        $r = "\n";
        $r .= "<div id='map$uuid' style='height:${height}px'></div>\n";
        $j .= "// start map $uuid\n";
        $j .= "jQuery(document).ready(function() {\n";
        $j .= "var map$uuid = new L.Map('map$uuid', osmAKuechlerConfig.mapConfig);\n";
        $j .= "map$uuid.attributionControl.setPrefix('');\n";
        $j .= "var baselayer$uuid = new L.TileLayer(osmAKuechlerConfig.server, osmAKuechlerConfig.tileConfig);\n";
        $j .= "var koord$uuid     = new L.LatLng($lat, $lon);\n";
        $j .= "var marker$uuid    = new L.Marker(koord$uuid);\n";
        $j .= "map$uuid.addLayer(marker$uuid);\n";
        $j .= "map$uuid.setView(koord$uuid, $zoom).addLayer(baselayer$uuid);\n";
        $j .= "marker$uuid.bindPopup('$popup');\n";
        $j .= "});\n";
        $j .= "// end map $uuid\n";
        $document->addScriptDeclaration ( $j );
        return $r;
    }

    function _getVariableOld($name, &$match) {
        $regex = '/' . $name . '=(?:\'(.+?)\'|([^\s]+))/i';
        preg_match ( $regex, $match, $matches );
        return (isset ( $matches [1] ) ? $matches [1] : '') . (isset ( $matches [2] ) ? $matches [2] : '');
    }
}
?>
